proposed/0000-Unified-structured-data.md

```diff
@@ -0,0 +1,128 @@
+on the network and how it is handled.
+
+###Data storage and retrieval
+
+ImmutableData is fixed self validating non mutable chunks. These require StructuredData types to manipulate information. These structured Data types may then create a global application acting on a key value store with very high degrees of availability and security (i.e. create network scale apps). Such apps could easily include medical condition analysis linked with genomic and proteomic sequencing to advance health based knowledge on a global scale. This proposal allows such systems to certainly be prototyped and tested with a high degree of flexibility.
+
+###New protocols
+
+As these data types are now self validating and may contain different information, such as new protocols, `rdf`/`owl` data types, the limit of new data types and ability to link such data is extremely scalable. Such protocols could indeed easily encompass token based systems (a form of 'crypto-currency'), linked data, natural language learning databases, pre-compilation units, distributed version control systems (git like) etc.
+
+###Compute
+
+Such a scheme would allow global computation types, possibly a Domain Specific Language (DSL) would define operator types to allow combination of functions. These could be made monotonic and allow out of order processing of programs (disorderly programming) which in itself presents an area that may prove to be well aligned with decentralised 'intelligence' efforts. Linked with 'zk-snarks' to alleviate any 'halting problem' type issues then a global Turing complete programming environment that optionally acts on semantic ('owl' / 'json-ld' etc.) data is a possible.
+
+##Expected outcome
+
+
+It is expected removing Transaction Managers from network will reduce complexity, code and increase security on the network, whilst allowing a greater degree of flexibility. These types now allow the users of this network to create their own data types and structures to be securely managed. This is hoped to allow many new types of application to exist.
+
+# Detailed design
+
+The design entails reducing all StructuredData types to a single type, therefore it should be able to be recognised by the network as StructuredData and all such sub-types handled exactly in the same manner. 
+
+##StructuredData
+
+```
+struct StructuredData {
+tag_type : TagType, // 8 Bytes ?
+identifier : NameType // 64Bytes (i.e. SHA512 Hash)
+data : mut Vec<u8>, // in many cases this is encrypted
+owner_keys : mut vec<crypto::sign::PublicKey> // n * 32 Bytes (where n is number of owners)
+version : mut u64, // incrementing (deterministic) version number
+previous_owner_keys : mut vec<crypto::sign::PublicKey> // n * 32 Bytes (where n is number of
+owners) only required when owners change 
+signature : mut Vec<Signature> // signs the `mut` fields above // 32 bytes (using e25519 sig)
+}
+```
+
```

Comment:
>  It could also be the scenario where only a single owner exists, if he sends two conflicting requests concurrently

I did not think of this one. I think this is a good reason to use vector clock for versioning the chunk (i.e. instead of `Vec<OwnerPubKey>` for owners we would use `Vec<(OwnerPubKey, VersionNumber)>`), that way we wouldn't rely on consensus which seems non deterministic anyway given the above. Instead we would gain the opportunity to deterministically tell whether the update took place or not.